### ISSUE TITLE
Literal types support in numpy.typing

### ISSUE THREAD

**Issue by zhihaoy** [numpy/numpy#16676]
Many NumPy APIs currently have return type `Any`.  `view`, `item`, `finfo`... But almost always, their type arguments are literals (`'d'`, `np.float64`, for example).  We should implement basic dependent types with Python 3.8 literal types in the typing module.

See also: https://www.python.org/dev/peps/pep-0586/

**Comment by person142** [comment]
Not sure what you mean here-`ndarray.view` does not have an `Any` return type in the types currently. For `ndarray.item`, it can return anything, i.e.

```
>>> class A: pass
...
>>> np.array([A()]).item()
<__main__.A object at 0x105ee6ed0>
```

The fix there is to make ndarray generic over the dtype, which is currently in progress. For `finfo`, nobody has added types yet; we have https://github.com/numpy/numpy/issues/16546 to track that.

**Comment by zhihaoy** [comment]
> Not sure what you mean here-`ndarray.view` does not have an `Any` return type in the types currently.

I mean `a.view('i8')`.

Yes, `item()` is different.  I mean all APIs that can accept `type` or `dtype` as function parameters.

**Comment by BvB93** [comment]
> I mean a.view('i8').

Right, so what you're talking about here boils down to [`DTypeLike`](https://github.com/numpy/numpy/blob/c37dc5f1cf5d42ddf05e0cd04a7fc528f30f20e5/numpy/typing/_dtype_like.py#L18) being (in practice) unable to accept arbitrary strings.

While aforementioned annotation should eventually be refined with Literals, 
there is already substantially sized todo-list (see the previously linked issue) with, as of yet, untyped functions.

**Comment by jorenham** [comment]
Progress has been made towards this in https://github.com/numpy/numpy/pull/26871 .
Nowadays there's also the private [`numpy._typing._char_codes`](https://github.com/numpy/numpy/blob/d60444f2383b9549e02bb61db956b91a5110ead1/numpy/_typing/_char_codes.py) module, which is used in e.g. the `numpy.dtype` constructor.

But I think that for 3rd party libraries it would indeed be helpful to expose (some of) these `Literal` aliases thorugh the public `numpy.typing` API. But since there are quite a few of them, we should take great care not to clutter the `numpy.typing` public namespace.
